Chapters 1 & 2 Revisions 05-02-2025 (#5113)
dfinity/motoko: doc/md/fundamentals/2-types/10-options.md

An option is defined using `?` followed by the type of the value it can hold.

```motoko name=user
var username : ?Text = null;
username;
```

`username` is an optional [`Text`](https://internetcomputer.org/docs/motoko/base/Text) value that starts as `null` (no username set).

:::info Null semantics

The constant `null` is the sole value of Motoko’s trivial `Null` type. It also represents the absence of a value in an optional type (`?T`). In this role, `null` is similar to `None` in Rust and Scala, and to `Nothing` in Haskell. Likewise, Motoko’s optional type `?T` is conceptually similar to `Option<T>` in Rust, `Option[T]` in Scala, and `Maybe T` in Haskell. In all these languages, the type system enforces explicit handling of missing values by representing optionality through a dedicated type.

:::

When a Motoko value has type `?T`, it is either `null` or contains a value, written as `?value` (note the leading `?`). The fundamental way to access this value is by using a `switch` expression, which explicitly handles both cases:

> ``` motoko include=user

Reviewer: this gives an error

> ``` motoko include=user

Reviewer: ```suggestion
``` motoko no-repl
```

func displayName(option : ?Text) : Text {
  switch option {
    case (?user) { user };
    case null { "Guest" };
  }
};
displayName(username);
```

Sometimes, the verbosity of a `switch` expression can make code harder to read. To improve readability, Motoko provides additional constructs for working with option types.

### Checking for presence

To determine if an option contains a value, function `Option.isSome` returns `true` if its argument is not `null`.

```motoko
import Option "mo:base/Option";
import Debug "mo:base/Debug";

let value : ?Nat = ?5;
if (Option.isSome(value)) {
  Debug.print("Value is present.");
}
```

By leveraging the `Option` module, handling optional values becomes more concise and expressive, reducing the need for explicit [`switch`](https://internetcomputer.org/docs/motoko/fundamentals/control-flow/switch) statements.

### Providing default values

Instead of manually handling `null` cases with [pattern matching](https://internetcomputer.org/docs/motoko/fundamentals/pattern-matching), `Option.get` allows for cleaner fallback logic to ensure that missing values are safely replaced with a default.

> ```motoko include=user

Reviewer: This also returns an error and does not compile in the preview

> ```motoko include=user

Reviewer: ```suggestion
```motoko no-repl
```

---
sidebar_position: 10
---

# Options

> | Type    | Syntax                  | Purpose                          | When to use it                                   |

Reviewer: ```suggestion
| Type    | Syntax                  | Purpose                          | Application                              |
```

|---------|-------------------------|----------------------------------|------------------------------------------------|
| Option  | `?T`                     | Represents a value that may be missing. | When an operation might return **no result**. |

Options provide a structured way to represent values that may or may not be present. Instead of using `null` directly, the option type enforces explicit handling, making programs safer and reducing unexpected failures.

An option is defined using `?` followed by the type of the value it can hold.

> ```motoko

Reviewer: does not return anything, should be no-repl.

> ```motoko

Reviewer: ```suggestion
```motoko no-repl
```

let username : ?Text = null;
Option.get(username, "Guest"); // "Guest" if username is null
```

### Using options for error handling

Options can be used to catch expected failures instead of calling a [`trap`](https://internetcomputer.org/docs/motoko/fundamentals/basic-syntax/traps), making a function return `null` when it encounters an invalid input.

```motoko
func safeDivide(a : Int, b : Int) : ?Int {
    if (b == 0) null else ?(a / b);
};

let result1 = safeDivide(10, 2); // ?5
safeDivide(10, 0); // null
```

Another way to extract values from option types is by using the `let ... else` pattern. This approach can be preferable to a `switch` expression for brevity and clarity. However, it only applies when the `else` branch can redirect control flow, such as returning early or throwing an error, if the value does not match the pattern in the `let`.

Another way to extract values from options, is use the `let ... else` pattern. This may be preferable over a `switch` expression, but only applies when the `else` expression can divert control elsewhere when the value fails to match the pattern used in the `let`.

For example, here’s a simple implementation of an option helper:

> ```motoko

Reviewer: does not compile in repl

```suggestion
```motoko no-repl
```

func get<T>(option : ?T, defaultValue : T) : T {
  let ?value = option else return defaultValue;
  return value;
};

assert get(?"A", "B") == "A";
assert get(null, "B") == "B";

> The `let` statement matches the option against the pattern `?value`, extracting the contained value if present. If the option is `null`, the pattern fails to match and the `else` branch is executed, typically to exit the function early or return a default value.

Reviewer: ````suggestion
```

The `let` statement matches the option against the pattern `?value`, extracting the contained value if present. If the option is `null`, the pattern fails to match and the `else` branch is executed, typically to exit the function early or return a default value.
````

The same logic can be expressed using a `switch`, though the result is more verbose and introduces an additional level of nesting:

``` motoko no-repl

> ``` motoko

Reviewer: ```suggestion
```

func get<T>(option : ?T, defaultValue : T) : T {
  switch option {
    case null defaultValue;
    case (?value) value;
  }
};
```
Although convenient for option patterns, `let-else` also works with other types of patterns.

> Although convenient for option patterns, `let-else` also works with other types of pattern,
> handling pattern-match failure using the `else` branch.
> :::

Reviewer: ```suggestion

```

```motoko
import Option "mo:base/Option";

let firstName : ?Text = ?"Motoko";
let lastName : ?Text = ?"Ghost";

func combineNames(f : Text, l : Text) : Text {
    f # " " # l;
};

Option.chain<Text, Text>(firstName, func (first : Text) {
    Option.map<Text, Text>(lastName, func(last : Text) { combineNames(first, last) });
});

// ?("Motoko Ghost")
```

## Option blocks (`do ?`)

Option blocks in Motoko use the syntax `do ? <block>` to work with optional values of type `?T` without requiring nested `switch` statements. When the `<block>` evaluates to a value of type `T`, the entire `do ?` expression returns a value of type `?T`. Crucially, it allows early exits from the block when encountering `null`.

Within a `do ? <block>`, the `!` operator is used to unwrap values of unrelated option types (e.g., `?U`). When evaluating an expression `<exp> !`, if `<exp>` results in `null`, control immediately exits the `do ?` block with value `null`. Otherwise, it unwraps the value `?v` and continues with `v` of type `U`.

The `do ? <block>` construct is similar to the `?` operator in Rust, providing a concise and expressive way to propagate `null` values.

> The `do ? <block>` is similar to how the `?` operator works in languages like Rust.

Reviewer: ```suggestion
```

```motoko 
 // Returns the sum of optional values `n` and `m` or `null`, if either is `null`
func addOpt(n : ?Nat, m : ?Nat) : ?Nat {
    do ? { 
      n! + m!  
    }
}

> 

Reviewer: ````suggestion
```
````

> The following example defines a simple function that evaluates expressions built from natural numbers, division and a zero test, encoded as a variant type:

Reviewer: ```suggestion
The following example defines a simple function that evaluates expressions built from natural numbers, division, and a zero test, encoded as a variant type:
```

---
sidebar_position: 10
---

# Options

| Type    | Syntax                  | Purpose                          | When to use it                                   |
|---------|-------------------------|----------------------------------|------------------------------------------------|
| Option  | `?T`                     | Represents a value that may be missing. | When an operation might return **no result**. |

Options provide a structured way to represent values that may or may not be present. Instead of using `null` directly, the option type enforces explicit handling, making programs safer and reducing unexpected failures.

An option is defined using `?` followed by the type of the value it can hold.

```motoko
var username : ?Text = null;
```

`username` is an optional [`Text`](https://internetcomputer.org/docs/motoko/base/Text) value that starts as `null` (no username set).

:::info Null semantics

> The constant `null` is the only value of the trivial `Null` type.  It also indicates an absent value in an optional type (`?T`). This makes `null` similar to `None` in languages like Rust, Scala, and `Nothing` in Haskell, and Motoko's option `?T` type similar to `Option<T>` in Rust, `Option[T]` in Scala, and `Maybe T` in Haskell.  In all these languages the type system enforces explicit handling of missing values by using a dedicated type to indicate optionality

Reviewer 2: ```suggestion
The constant `null` is the only value of the trivial `Null` type.  It also indicates an absent value in an optional type (`?T`). This makes `null` similar to `None` in languages like Rust, Scala, and `Nothing` in Haskell, and Motoko's option `?T` type similar to `Option<T>` in Rust, `Option[T]` in Scala, and `Maybe T` in Haskell.  In all these languages the type system enforces explicit handling of missing values by using a dedicated type to indicate optionality.
```

:::

> When a Motoko value has type `?T`, then is either `null` or some value, written `?value` (note leading `?`). The fundamental way to access the value is to use a `switch` expression, that explicitly handles both cases:

Reviewer 2: ```suggestion
When a Motoko value has type `?T`, then is either `null` or some value, written `?value` (note the leading `?`). The fundamental way to access the value is to use a `switch` expression, that explicitly handles both cases:
```

``` motoko
func displayName(option : ?Text) : Text {
  switch option {

>      case (?user) { user };
>      case (null) { "Guest" };

Reviewer 2: ```suggestion
     case ?user { user };
     case null { "Guest" };
```
I'm pretty sure the parens are unnecessary...(but are needed for variants)

  }
};
displayName(null);
```

Sometimes, the verbosity of `switch` expression can make code harder to read, so Motoko also offers additional ways to handle option types.

### Checking for presence

> To determine if an option contains a value, `Option.isSome` returns `true` if it is not `null`.

Reviewer 2: ```suggestion
To determine if an option contains a value, function `Option.isSome` returns `true` if its argument is not `null`.
```

let value : ?Nat = ?5;
if (Option.isSome(value)) {
    Debug.print("Value is present.");
}
```

By leveraging the `Option` module, handling optional values becomes more concise and expressive, reducing the need for explicit [`switch`](https://internetcomputer.org/docs/motoko/fundamentals/control-flow/switch) statements.

### Providing default values

Instead of manually handling `null` cases with [pattern matching](https://internetcomputer.org/docs/motoko/fundamentals/pattern-matching), `Option.get` allows for cleaner fallback logic to ensure that missing values are safely replaced with a default.

```motoko
import Option "mo:base/Option";

let username : ?Text = null;
Option.get(username, "Guest"); // "Guest" if username is null
```

### Using options for error handling

Options can be used to catch expected failures instead of calling a [`trap`](https://internetcomputer.org/docs/motoko/fundamentals/basic-syntax/traps), making a function return `null` when it encounters an invalid input.

```motoko
func safeDivide(a : Int, b : Int) : ?Int {
    if (b == 0) null else ?(a / b);
};

let result1 = safeDivide(10, 2); // ?5
safeDivide(10, 0); // null
```

### Let / else

> Another way to extract values from options, is use the `let ... else` pattern. This may be preferable over a `switch` expression, but only applies when the `else` expression can divert control elsewhere when the value fails to match the pattern used in the `let`

Reviewer 2: ```suggestion
Another way to extract values from options, is use the `let ... else` pattern. This may be preferable over a `switch` expression, but only applies when the `else` expression can divert control elsewhere when the value fails to match the pattern used in the `let`.
```

```motoko
import Option "mo:base/Option";

let firstName : ?Text = ?"Motoko";
let lastName : ?Text = ?"Ghost";

func combineNames(f : Text, l : Text) : Text {
    f # " " # l;
};

Option.chain<Text, Text>(firstName, func (first : Text) {
    Option.map<Text, Text>(lastName, func(last : Text) { combineNames(first, last) });
});

// ?("Motoko Ghost")
```

If either `firstName` or `lastName` is `null`, the result remains `null`.

## Option blocks (`do ?`)

Option blocks use the syntax `do ? <block>` to handle optional values of type `?T` without needing nested switch statements. It produces a value of type `?T`, when `<block>` has type `T` and, importantly, introduces the possibility of a break from `<block>.`

Within a `do ? <block>`, the `null` break `<exp> !` tests whether the result of the expression, `<exp>` of unrelated option type, `?U`, is `null`.

If the result is `null`, control immediately exits the `do ? <block>` with value `null`. Otherwise, the result of `<exp>` must be an option value `?v`, and evaluation of `<exp> !` proceeds with its contents, `v` of type `U`.

The `do ? <block>` is similar to how the `?` operator works in languages like Rust.

> ```motoko no-repl
> // Introduces an option block that returns a value of type ?T
> let result: ?T = do ? {
>   // The ! operator (null break) unwraps optional values inside the block
>   someOptionalValue!  // unwraps the value or short-circuits
> }
> ```

Reviewer 2: Your example isn't wrong, but also doesn't do anything very interesting apart from unwrap and re-wrap an option.
Maybe this is better?
```suggestion
```motoko 
 // Returns the sum of optional values `x` and `y` or `null`, if either is `null`
func addOpt(x:?Nat, y:?Nat) : ?Nat {
    do ? { 
      x! + y!  
    }
}
```

The following example defines a simple function that evaluates expressions built from natural numbers, division and a zero test, encoded as a variant type:

```motoko
type Exp = {#Lit : Nat; #Div : (Exp, Exp); #If : (Exp, Exp, Exp)};
func eval(e : Exp) : ? Nat {
  do ? {
    switch e {
      case (#Lit n) { n };
      case (#Div (e1, e2)) {
        let v1 = eval e1 !;  // If eval e1 returns null, exit with null
        let v2 = eval e2 !;  // If eval e2 returns null, exit with null
        if (v2 == 0)
          null !  // Explicitly exit with null for division by zero
        else v1 / v2
      };
      case (#If (e1, e2, e3)) {
        if (eval e1 ! == 0)  // Unwrap and check if zero
          eval e2 !  // Return result of e2 (or null if it's null)
        else
          eval e3 !  // Return result of e3 (or null if it's null)
      };
    };
  };
};

let expr : Exp = #If(

>   #Div(#Lit 10, #Lit 2),   // 10 / 2 = 5 (non-zero, so evaluate e3)

Reviewer 2: (this suggestion would replace a line of the code, breaking the code).

  #Lit 0,                  // e2 (ignored because e1 ≠ 0)
  #Div(#Lit 6, #Lit 3)     // e3 → 6 / 3 = 2
);

eval(expr);
```

To guard against division by 0 without trapping, the eval function returns an option result, using `null` to indicate failure.

> Each recursive call is checked for `null` using `!`, immediately exiting the outer `do ?` block, and then the function itself, when a result is `null`.

Reviewer 2: ```suggestion
Each recursive call is checked for `null` using `!`, immediately exiting the outer `do ?` block, and then the function itself, when an intermediate result is `null`.
```
